File: packages/manufacturing-types/src/types/cad.ts
import { z } from 'zod';

export const CADFileFormatSchema = z.enum([
  'step',
  'iges',
  'stl',
  'obj',
  'fbx',
  'gltf',
  'dwg',
  'dxf',
  'parasolid',
  'stp',
  'sldprt',
  'iam',
  'prt',
]);

export type CADFileFormat = z.infer<typeof CADFileFormatSchema>;

export const BoundingBoxSchema = z.object({
  min_x: z.number(),
  min_y: z.number(),
  min_z: z.number(),
  max_x: z.number(),
  max_y: z.number(),
  max_z: z.number(),
});

export type BoundingBox = z.infer<typeof BoundingBoxSchema>;

export const CADModelSchema = z.object({
  id: z.string(),
  name: z.string(),
  format: CADFileFormatSchema,
  file_url: z.string(),
  file_hash: z.string(),
  file_size_bytes: z.number(),
  bounding_box: BoundingBoxSchema,
  volume_cubic_mm: z.number().optional(),
  surface_area_mm2: z.number().optional(),
  part_count: z.number().optional(),
  assembly_count: z.number().optional(),
  properties: z.record(z.string(), z.unknown()).optional(),
  created_at: z.date(),
  updated_at: z.date(),
});

export type CADModel = z.infer<typeof CADModelSchema>;

export const GeometryFeatureSchema = z.object({
  id: z.string(),
  name: z.string(),
  type: z.enum(['hole', 'pocket', 'boss', 'fillet', 'chamfer', 'thread', 'groove', 'edge', 'face', 'surface']),
  dimensions: z.record(z.string(), z.number()),
  properties: z.record(z.string(), z.unknown()).optional(),
});

export type GeometryFeature = z.infer<typeof GeometryFeatureSchema>;

export const CADAnalysisSchema = z.object({
  id: z.string(),
  cad_model_id: z.string(),
  analysis_type: z.enum(['geometry', 'draft', 'wall_thickness', 'draft_angle', 'undercut', 'sharp_edges', 'surface_continuity']),
  features: z.array(GeometryFeatureSchema),
  issues: z.array(z.object({
    issue_id: z.string(),
    type: z.string(),
    severity: z.enum(['info', 'warning', 'error', 'critical']),
    description: z.string(),
    location: z.object({
      x: z.number(),
      y: z.number(),
      z: z.number(),
    }).optional(),
    recommendation: z.string().optional(),
  })),
  manufacturability_score: z.number().min(0).max(100),
  created_at: z.date(),
  updated_at: z.date(),
});

export type CADAnalysis = z.infer<typeof CADAnalysisSchema>;

export const CADRevisionSchema = z.object({
  id: z.string(),
  cad_model_id: z.string(),
  revision_number: z.number(),
  changes: z.array(z.object({
    field: z.string(),
    old_value: z.unknown(),
    new_value: z.unknown(),
  })),
  change_reason: z.string().optional(),
  created_by: z.string(),
  created_at: z.date(),
});

export type CADRevision = z.infer<typeof CADRevisionSchema>;

File: apps/api/app/schemas/cad.py
"""CAD and analysis schemas"""

from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime
from decimal import Decimal
from typing import Optional, List, Dict, Any


class BoundingBox(BaseModel):
    min_x: float
    min_y: float
    min_z: float
    max_x: float
    max_y: float
    max_z: float


class CADModelBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    format: str = Field(..., min_length=1, max_length=20)
    file_url: str = Field(..., max_length=500)
    file_hash: Optional[str] = Field(None, max_length=256)
    file_size_bytes: Optional[int] = None
    volume_cubic_mm: Optional[float] = None
    surface_area_mm2: Optional[float] = None
    part_count: Optional[int] = None
    assembly_count: Optional[int] = None


class CADModelCreate(CADModelBase):
    hardware_part_id: Optional[UUID] = None
    bounding_box: Optional[BoundingBox] = None


class CADModelRead(CADModelBase):
    id: UUID
    hardware_part_id: Optional[UUID] = None
    bounding_box_min_x: Optional[float] = None
    bounding_box_min_y: Optional[float] = None
    bounding_box_min_z: Optional[float] = None
    bounding_box_max_x: Optional[float] = None
    bounding_box_max_y: Optional[float] = None
    bounding_box_max_z: Optional[float] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class CADAnalysisBase(BaseModel):
    analysis_type: Optional[str] = None
    manufacturability_score: int = Field(50, ge=0, le=100)
    has_issues: bool = False
    issues_count: int = 0
    dfm_score: Optional[int] = None
    estimated_machining_time_minutes: Optional[int] = None
    estimated_cost: Optional[Decimal] = Field(None, decimal_places=2)
    estimated_lead_time_days: Optional[int] = None
    features: List[Dict[str, Any]] = []
    issues: List[Dict[str, Any]] = []
    recommendations: List[Dict[str, Any]] = []


class CADAnalysisCreate(CADAnalysisBase):
    hardware_part_id: Optional[UUID] = None
    cad_model_id: Optional[UUID] = None


class CADAnalysisRead(CADAnalysisBase):
    id: UUID
    hardware_part_id: Optional[UUID] = None
    cad_model_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class CADAnalysisUpdate(BaseModel):
    analysis_type: Optional[str] = None
    manufacturability_score: Optional[int] = None
    has_issues: Optional[bool] = None
    issues_count: Optional[int] = None
    dfm_score: Optional[int] = None
    estimated_machining_time_minutes: Optional[int] = None
    estimated_cost: Optional[Decimal] = None
    estimated_lead_time_days: Optional[int] = None
    features: Optional[List[Dict[str, Any]]] = None
    issues: Optional[List[Dict[str, Any]]] = None
    recommendations: Optional[List[Dict[str, Any]]] = None
